### PR TITLE
[Fix] Patching nncf_model_inputs

### PR DESCRIPTION
**Pull request by alexsu52**
### Changes
Strip traced tensors when exit from the tracing context

### Reason for changes

`nncf_model_input` patched `torch.Tensor` with `TracedTensor` which does not support deepcopy.

### Related tickets

125357

### Tests
test_wrap_model_with_example_input
e2e 506
test_examples 154

### CODE REVIEW

**Review by vshampor — CHANGES_REQUESTED** [review]
I have proposed one possible solution below, but as I have been writing this, I have thought of another one. 

We could try and keep a weakref to `self` in the `TracedTensor`, and also preserve weakrefs to input tensors inside the `self.nncf._compressed_context`  somewhere around the `args, kwargs = self.nncf._wrap_inputs_fn(args, kwargs)` line in `NNCFNetwork.forward`. We could then in `nncf.torch.dynamic_graph.context.TracingContext.__exit__` iterate over these "input tensor weakrefs", see if they are still valid (they probably will be valid, but just in case), and if so, replace the `.__class__` attribute on the corresponding tensors back to `torch.Tensor` (or to a previously saved "original class"). 

In this way, upon exiting NNCFNetwork all `TracedTensors` will get automatically turned back into regular `Tensors`. There may be issues with copyability/shallow copies of the `TracedTensor` while it still has a weakref, though, would have to test for that first.

On the other hand, the complexity of this solution is moderate, and it would indeed be easier just to do the torch.Tensor copy for the time being, and hope that the users will forgive us the extra input tensor copies. If you think that the perf overhead will not be an issue in practice , then I'll approve this if you only address the comment about the test case.

**Review by vshampor — DISMISSED** [review]
Great job cleaning out the hidden issues in the context. I see the new test on nested contexts and tensor tracing, but still consider adding the test that would fail with a `__deepcopy__` related error we've been seeing in PTQ tests without the fix, and would *not* fail with the fix, so that we catch this kind of errors early on (e.g. at precommit stage).

File: nncf/torch/dynamic_graph/context.py [hunk 1]
```diff
@@ -122,25 +123,29 @@ def __enter__(self):
...
-            for traced_tensor_weakref in self._threading.thread_local.traced_tensor_weakrefs:
-                tt = traced_tensor_weakref()
-                if tt is not None:
-                    tt.nncf_expire()
-
-        self._reset_thread_local()
-
-        set_current_context(self._save_context)
-        self._save_context = None
+        save_context = self._threading.thread_local.save_context.pop(-1)
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/torch/dynamic_graph/context.py:132]: done

File: tests/torch/test_tracing_context.py [hunk 2]
```diff
@@ -118,3 +118,22 @@ def test_no_cross_forward_run_dependency():
...
+            with ctx:
+                module(tensor)
+                assert len(ctx._threading.thread_local.save_context) == 3
+                assert len(ctx._threading.thread_local.traced_tensor_weakrefs) > 0
+            assert len(ctx._threading.thread_local.save_context) == 2
+            assert len(ctx._threading.thread_local.traced_tensor_weakrefs) > 0
+        assert len(ctx._threading.thread_local.save_context) == 1
+        assert ctx._threading.thread_local.save_context[0] is None
+        assert len(ctx._threading.thread_local.traced_tensor_weakrefs) > 0
+    assert len(ctx._threading.thread_local.traced_tensor_weakrefs) == 0
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply tests/torch/test_tracing_context.py:139]: Thanks for the comment. Done.

File: nncf/torch/dynamic_graph/trace_tensor.py [hunk 3]
```diff
@@ -166,14 +157,30 @@ def make_tensor_metas(inputs: OperatorInput) -> List[Optional[TensorMeta]]:
...
     return tensor_metas
+
+
+def strip_traced_tensor(tensor: TracedTensor) -> torch.Tensor:
+    """
+    Returns the tensor object with tracing additions removed.
+    """
+    tensor.__class__ = tensor.original_class
+    return tensor
+
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/torch/dynamic_graph/trace_tensor.py:175]: Could you provide your opinion about the method signature?

**Comment by alexsu52** [comment]
> Great job cleaning out the hidden issues in the context. I see the new test on nested contexts and tensor tracing, but still consider adding the test that would fail with a `__deepcopy__` related error we've been seeing in PTQ tests without the fix, and would _not_ fail with the fix, so that we catch this kind of errors early on (e.g. at precommit stage).

The test has been added.

File: nncf/torch/dynamic_graph/trace_tensor.py [hunk 4]
```diff
@@ -57,24 +58,14 @@ class TracedTensor(torch.Tensor):
     """
 
     @staticmethod
-    def from_torch_tensor(tensor, tensor_meta: TensorMeta):
+    def from_torch_tensor(tensor: torch.Tensor, tensor_meta: TensorMeta) -> "TracedTensor":
         tensor.tensor_meta = tensor_meta
```
Comment (AlexanderDokuchaev) [inline nncf/torch/dynamic_graph/trace_tensor.py:62]: Could you add docstring?

**Review by AlexanderDokuchaev — CHANGES_REQUESTED** [review]

File: nncf/torch/dynamic_graph/trace_tensor.py [hunk 5]
```diff
@@ -166,14 +157,30 @@ def make_tensor_metas(inputs: OperatorInput) -> List[Optional[TensorMeta]]:
...
         else:
             tensor_metas.append(None)
     return tensor_metas
+
+
+def strip_traced_tensor(tensor: TracedTensor) -> torch.Tensor:
+    """
+    Returns the tensor object with tracing additions removed.
+    """
+    tensor.__class__ = tensor.original_class
```
[the remark this answers is not in the document]
Comment (alexsu52) [reply nncf/torch/dynamic_graph/trace_tensor.py]: done

File: nncf/torch/dynamic_graph/trace_tensor.py (hunk 4 above)
Comment (alexsu52) [reply nncf/torch/dynamic_graph/trace_tensor.py:69]: done

**Review by AlexanderDokuchaev — APPROVED** [review]

**Pull request merged**